Add new field takeintoaccountdate to the glpi_tickets

VladoTTX opened the pull request:
Add a new field takeintoaccountdate to the glpi_tickets to fix existing issues caused by the usage of takeintoaccount_delay_stat.
New field takeintoaccountdate is updated with timestamp when Take Into Account (TIA) has happened.
This will fix issues with displaying and computation of proper TIA and also will help reporting to calculate whether TTO has been met or no, as it will be simple comparison of two values.
Migration script is provided which will compute and store approximate time when Take Into Account has happened on the existing tickets.

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #12167 #11109

cedric-anne commented on src/Ticket.php:
```diff
@@ -1529,10 +1529,19 @@ public function pre_updateInDB()
+	if(in_array("takeintoaccount_delay_stat",$this->updates) &&
+            $this->fields['takeintoaccount_delay_stat'] == 0
+	) {
+	    !in_array("takeintoaccountdate",$this->updates) && $this->updates[] = "takeintoaccountdate";
```
```suggestion
	    if (!in_array("takeintoaccountdate",$this->updates)) {
	        $this->updates[] = "takeintoaccountdate";
	    }
```

cedric-anne commented on src/Search.php:
```diff
@@ -6193,13 +6193,13 @@ public static function displayConfigItem($itemtype, $ID, $data = [])
                 $status       = $data[$NAME][0]['status'];
                 $opening_date = $data[$NAME][0]['date'];
-                $tia_time     = $data[$NAME][0]['takeintoaccount_delay_stat'];
+                $tia_date     = $data[$NAME][0]['takeintoaccountdate'];
```
This will require to replace `takeintoaccount_delay_stat` by `takeintoaccountdate` in `additionalfields` of corresponding search options.

VladoTTX replied:
thanks fixed by https://github.com/glpi-project/glpi/pull/12613/commits/24a029ea8a20c111a0097596e654af74edb8e72b

cedric-anne commented on install/migrations/update_10.0.2_to_10.0.3/ticket.php:
```diff
@@ -0,0 +1,74 @@
+    $migration->addKey("glpi_tickets", "takeintoaccountdate");
+    $migration->migrationOneTable("glpi_tickets");
+
+    foreach (getAllDataFromTable('glpi_tickets', ['takeintoaccount_delay_stat' => ['>', 0]]) as $data)
+    {
```
It would be safer to not query all tickets data.
```suggestion
    $tickets_iterator = $DB->request([
        'SELECT'    => ['id'],
        'FROM'      => 'glpi_tickets',
        'WHERE'     => [
            'takeintoaccount_delay_stat' => ['>', 0],
        ]
    ]);

    foreach ($tickets_iterator as $data) {
```

cedric-anne reviewed (commented):
Recomputing take into account date for old tickets may produce unexpected results, as calendar may have be reconfigured.
We have 2 choices here:
1. Keep a null value, and in case value is null but `takeintoaccount_delay_stat` is not, we fallback to the previous logic, which was producing wrong results. We can eventually add a warning to the user, in front of ticket stats, to explain this known issue.
2. Recompute values, and accept that sometimes it will not correspond to the real date.

I personally prefer the first solution.

cedric-anne commented on install/migrations/update_10.0.2_to_10.0.3/ticket.php:
```diff
@@ -0,0 +1,81 @@
+    foreach ($tickets_iterator as $data) {
+       $ticket = new Ticket();
+       $ticket->getFromDB($data['id']);
+       $tia_delay = $ticket->getField('takeintoaccount_delay_stat');
+       $ticket_date = $ticket->getField('date');
```
`getField` is for display purpose only, as it may return `N/A`.
```suggestion
       $tia_delay = $ticket->fields['takeintoaccount_delay_stat'];
       $ticket_date = $ticket->fields['date'];
```

VladoTTX commented:
Allowing NULL values would complicate the reporting and also other SQL queries example generateSLAOLAComputation().

You still will need to recomp]ute the TIA. Only difference would be that it in 1st case it will be done always during displaying/query and 2nd case it would be done just once during migration.

Also handling of TIA is incorrect and should be fixed..
Ticket::computeTakeIntoAccountDelayStat() is using Ticket::getCalendar() which now uses SLA::TTR calendar.
This does not count with situation that TTO can have different calendar. In that case it produces incorrect delay time which is already stored as incorrect in DB.

IHMO we should migrate takeintoaccount_delay_stat to takeintoaccountdate (even inaccurately) at one point and change the computation to use TTO together with introduction of this field (eg.  fix Ticket::computeTakeIntoAccountDelayStat() to use TTO calendar SLA or OLA).

Still there is question what to do in cases when both SLA and OLA TTOs are defined.. 
To properly fix that it would probably require separating TIA for OLA (or reworking it completely) and I have no idea whether there is any interest in that..

cedric-anne commented:
Maybe another field, like `is_takeintoaccountdate_imprecise`, could be used as a flag and set to `1` by migration. With this, we would be able to add a :warning: icon where data is displayed. By default, this field value would be `0`, so only migration process would have to define it explicitely.

VladoTTX commented:
> Maybe another field, like `is_takeintoaccountdate_imprecise`, could be used as a flag and set to `1` by migration. With this, we would be able to add a ⚠️ icon where data is displayed. By default, this field value would be `0`, so only migration process would have to define it explicitely.

Not sure whether it is worth having additional fields just for this and keeping it for the future..
Maybe just store last ticket ID of ticket which was migrated/computed with old method..

orthagh commented on install/migrations/update_10.0.2_to_10.0.3/ticket.php:
```diff
@@ -0,0 +1,81 @@
+          // Compute takeintoaccountdate using Calendar, delay is added to active time of calendar
+          // this is just an approximation as time during non-active hours was not counted to delay (should be fixed with introduction of new field)
+          // Also if TIA happend before active hours of calendar it is 1 second as 0 has meaning "not taken into account yet"
+          $tia_update = $calendar->computeEndDate($ticket_date,$tia_delay,0,false,false);
```
@VladoTTX how many tickets do you have on your side to compute.
This part is very concerning in term of execuction time.

Let's think about instances with like 100k tickets (no so rare), if the computeEndDate take 100ms, we talk about 3hours of migration

cedric-anne replied:
IMHO, we should probably just create this field using a `NULL` value for existing tickets, and fallback to current computation when this value `NULL`.

VladoTTX replied:
You are right, this might take some time..
Personally I've tested this on very small instance with approx. of 1000 tickets.
I can do some measurements in larger data sets..

orthagh commented on install/migrations/update_10.0.2_to_10.0.3/ticket.php:
```diff
@@ -0,0 +1,81 @@
+          $tia_update  = date('Y-m-d H:i:s', strtotime($ticket_date) + $tia_delay);
+       }
+    
+       $query = "UPDATE `glpi_tickets` SET takeintoaccountdate = '" . $tia_update . "' WHERE id=" . $data['id'];
```
apart the above comment (which is the most important), maybe this query can be moved within a prepared statement executed outside the loop

VladoTTX replied:
Does not seem to be too much a performance intensive.. 
I've tested the code on set of 40k tickets on VM (4vCPU/8GB RAM) and it took 100 seconds to process without UPDATE statement and 150 seconds together with UPDATE.
Of course this could depend how big was TIA but in real life it would probably be small value..

why it would be better to have prepare/execute statement outside the loop?

cedric-anne commented on src/Ticket.php:
```diff
@@ -1547,9 +1558,22 @@ public function computeTakeIntoAccountDelayStat()
+            $calendars_id = 0;
             $calendar     = new Calendar();
-
+	    // Not clear how to handle case with both SLA and OLA
+	    // For now prefer OLA calendar
+	    if (isset($this->fields['olas_id_tto']) && $this->fields['olas_id_tto'] > 0){
+		    $la = new OLA();
+		    $la->getFromDB($this->fields['olas_id_tto']);
+	    } elseif (isset($this->fields['slas_id_tto']) && $this->fields['slas_id_tto'] > 0){
+		    $la = new SLA();
+		    $la->getFromDB($this->fields['slas_id_tto']);
+	    }
+	    if (isset($la) && $la->fields['use_ticket_calendar']){
+		    $calendars_id = $la->getField('calendars_id');
+	    } else {
+		    $calendars_id = parent::getCalendar();
+	    }
```
`takeintoaccount_delay_stat` reflects the "customer" point of view, and it should not be changed.
We have already identified that this internal "take into account" delay information is missing, but it should probably be store in a new `internal_takeintoaccount_delay_stat` field. Also, this information may be sensitive, from a commercial point of view, so I do not know how and where it should be displayed.

This should be changed in another PR, to keep small PRs that would be easier to validate.

VladoTTX replied:
Please check below, I've updated code to use just SLA TTO calendar (if defined) and fallback to previous call of Ticket::getCalendar() if it is not.

VladoTTX replied:
Ok, I've reverted the fixes to computeTakeIntoAccountDelayStat
I wanted to combine it, to have takeintoaccount_delay_stat correct after this PR (we have a case when SLA::TTO/TTR are using different calendars with different working hours)
I'll open a separate PR for that

VladoTTX replied:
Submitted PR #12646 for fixing computeTakeIntoAccountDelayStat

VladoTTX commented:
Guys, I've updated PR for 10.0.4 could you please check?
Also I've updated script to use prepared statement but it does not make much difference in speed for me.
On 40k tickets without prepared statement it took 150 seconds, now it was 140 seconds.

cedric-anne reviewed (changes requested):
We internally discuss about this with @orthagh and we agreed that we should not try to compute the `takeintoaccountdate` on ticket that were already taken into account.
Indeed, as calendar may have change since ticket creation, computed value may be wrong and we do not wan to store a wrong value in database. Also, we prefer not to introduce a costly migration on a bugfix version.

It means that:
 - this field value must stay `NULL` for existing tickets;
 - conditions should fallback to existing check on `takeintoaccount_delay_stat` when `takeintoaccountdate` is `NULL`, for instance `('takeintoaccountdate' IS NOT NULL AND ...) OR ('takeintoaccountdate' IS NULL AND ...)`.

It also we would be able to easilly identify tickets on which this date cannot be computed in an accurate way (if `takeintoaccount_delay_stat` is `> 0` and `takeintoaccountdate` is `null`), and so we will be able to add warnings about it on UI.

VladoTTX commented:
If you changed calendar you already (probably) have wrong value of takeintoaccount_delay_stat in DB, as it was computed using the old calendar.
And I think, more clean would be to migrate to takeintoaccountdate in the best way it could be done, instead of maintaining false sense of correctness.
Having both conditions will make reporting even more complicated as before. 
With takeintoaccountdate it would be just comparison, now both conditions needs to be maintained (one correct, one estimate)
I will try to update the PR but will try to think about another way as IMHO it is not best way to go..

VladoTTX commented:
I've tried to do the updates to count with NULL on old tickets.
Question is whether I should remove migration script or it could be made optional as part example glpi:migration:takeintoaccountdate

cedric-anne commented on src/CommonITILObject.php:
```diff
@@ -4291,10 +4292,13 @@ public static function generateSLAOLAComputation($type, $table = "TABLE")
                 return 'IF(' . $DB->quoteName($table . '.' . $type) . ' IS NOT NULL
             AND ' . $DB->quoteName($table . '.status') . ' <> ' . self::WAITING . '
-            AND (' . $DB->quoteName($table . '.takeintoaccount_delay_stat') . '
+            AND ((' . $DB->quoteName($table . '.takeintoaccountdelaydate') . ' IS NOT NULL AND 
```
```suggestion
            AND ((' . $DB->quoteName($table . '.takeintoaccountdate') . ' IS NOT NULL AND 
```

cedric-anne commented on src/CommonITILObject.php:
```diff
@@ -4291,10 +4292,13 @@ public static function generateSLAOLAComputation($type, $table = "TABLE")
-            AND (' . $DB->quoteName($table . '.takeintoaccount_delay_stat') . '
+            AND ((' . $DB->quoteName($table . '.takeintoaccountdelaydate') . ' IS NOT NULL AND 
+                 ' . $DB->quoteName($table . '.takeintoaccountdate') . ' > ' . $DB->quoteName($table . '.' . $type) . ')
+                 OR (' . $DB->quoteName($table . '.takeintoaccountdelaydate') . ' IS NULL AND
```
```suggestion
                 OR (' . $DB->quoteName($table . '.takeintoaccountdate') . ' IS NULL AND
```

cedric-anne commented on src/Ticket.php:
```diff
@@ -6220,10 +6232,15 @@ public function showStatsDates()
         $now                      = time();
         $date_creation            = strtotime($this->fields['date'] ?? '');
+       // Tickets created before 10.0.4 do not have takeintoaccountdate field, use old and incorrect computation for those cases
         $date_takeintoaccount     = $date_creation + $this->fields['takeintoaccount_delay_stat'];
         if ($date_takeintoaccount == $date_creation) {
             $date_takeintoaccount  = 0;
         }
+       // If takeintoaccountdate was saved use it
+        if(!empty($this->fields['takeintoaccountdate'])) {
+            $date_takeintoaccount = strtotime($this->fields['takeintoaccountdate']);
+        }
```
It would be easier to read, IMHO.
```suggestion
        $date_takeintoaccount     = 0;
        if ($this->fields['takeintoaccountdate'] !== null) {
            $date_takeintoaccount = strtotime($this->fields['takeintoaccountdate']);
        } elseif ($this->fields['takeintoaccount_delay_stat'] > 0) {
            $date_takeintoaccount = $date_creation + $this->fields['takeintoaccount_delay_stat'];
        }
```

cedric-anne commented:
> I've tried to do the updates to count with NULL on old tickets. Question is whether I should remove migration script or it could be made optional as part example glpi:migration:takeintoaccountdate

You should remove it from the PR.

VladoTTX commented:
> You should remove it from the PR.
Ok, removed the recompute part from the script
